File: listacirculardoble/listaCirculardoble.java
package estructura_de_datos.listacirculardoble;

import java.util.Scanner;

class Nodo {
    int dato;
    Nodo sig;
    Nodo ant;

    Nodo(int d){
        dato = d;
        sig = ant = null;
    }
}

public class listaCirculardoble {
    Nodo cabeza = null;

    // ============================================
    // INSERTAR AL INICIO
    // ============================================
    void insertarInicio(int dato){
        Nodo nuevo = new Nodo(dato);

        if (cabeza == null){
            cabeza = nuevo;
            nuevo.sig = nuevo;
            nuevo.ant = nuevo;
            return;
        }

        Nodo ultimo = cabeza.ant;

        nuevo.sig = cabeza;
        nuevo.ant = ultimo;

        ultimo.sig = nuevo;
        cabeza.ant = nuevo;

        cabeza = nuevo;
    }

    // ============================================
    // INSERTAR AL FINAL
    // ============================================
    void insertarFinal(int dato){
        Nodo nuevo = new Nodo(dato);

        if (cabeza == null){
            cabeza = nuevo;
            nuevo.sig = nuevo;
            nuevo.ant = nuevo;
            return;
        }

        Nodo ultimo = cabeza.ant;

        ultimo.sig = nuevo;
        nuevo.ant = ultimo;

        nuevo.sig = cabeza;
        cabeza.ant = nuevo;
    }

    // ============================================
    // INSERTAR EN POSICIÓN
    // ============================================
    void insertarPosicion(int dato, int pos){
        if (pos == 0){
            insertarInicio(dato);
            return;
        }

        if (cabeza == null){
            System.out.println("La lista está vacía");
            return;
        }

        Nodo temp = cabeza;
        for (int i = 0; i < pos - 1 && temp.sig != cabeza; i++)
            temp = temp.sig;

        Nodo nuevo = new Nodo(dato);
        Nodo siguiente = temp.sig;

        nuevo.sig = siguiente;
        nuevo.ant = temp;

        temp.sig = nuevo;
        siguiente.ant = nuevo;
    }

    // ============================================
    // ELIMINAR AL INICIO
    // ============================================
    void eliminarInicio(){
        if (cabeza == null) return;

        if (cabeza.sig == cabeza){
            cabeza = null;
            return;
        }

        Nodo ultimo = cabeza.ant;
        cabeza = cabeza.sig;

        cabeza.ant = ultimo;
        ultimo.sig = cabeza;
    }

    // ============================================
    // ELIMINAR AL FINAL
    // ============================================
    void eliminarFinal(){
        if (cabeza == null) return;

        if (cabeza.sig == cabeza){
            cabeza = null;
            return;
        }

        Nodo ultimo = cabeza.ant;
        Nodo nuevoUltimo = ultimo.ant;

        nuevoUltimo.sig = cabeza;
        cabeza.ant = nuevoUltimo;
    }

    // ============================================
    // ELIMINAR EN POSICIÓN
    // ============================================
    void eliminarPosicion(int pos){
        if (cabeza == null) return;

        if (pos == 0){
            eliminarInicio();
            return;
        }

        Nodo temp = cabeza;

        for (int i = 0; i < pos && temp.sig != cabeza; i++)
            temp = temp.sig;

        Nodo anterior = temp.ant;
        Nodo siguiente = temp.sig;

        anterior.sig = siguiente;
        siguiente.ant = anterior;

        // Si eliminan la cabeza real
        if (temp == cabeza)
            cabeza = siguiente;
    }

    // ============================================
    // MOSTRAR LISTA
    // ============================================
    void mostrar(){
        if (cabeza == null){
            System.out.println("Lista vacía");
            return;
        }

        Nodo temp = cabeza;

        do {
            System.out.print(temp.dato + " <-> ");
            temp = temp.sig;
        } while (temp != cabeza);

        System.out.println("(regresa)");
    }

    // ============================================
    // MAIN CON MENÚ
    // ============================================
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        listaCirculardoble lista = new listaCirculardoble();

        int opcion;

        do {
            System.out.println("\n=== MENÚ LISTA CIRCULAR DOBLE ===");
            System.out.println("1. Insertar al inicio");
            System.out.println("2. Insertar al final");
            System.out.println("3. Insertar en posición");
            System.out.println("4. Eliminar al inicio");
            System.out.println("5. Eliminar al final");
            System.out.println("6. Eliminar en posición");
            System.out.println("7. Mostrar lista");
            System.out.println("8. Salir");
            System.out.print("Opción: ");
            opcion = sc.nextInt();

            switch(opcion){
                case 1:
                    System.out.print("Dato: ");
                    lista.insertarInicio(sc.nextInt());
                    break;

                case 2:
                    System.out.print("Dato: ");
                    lista.insertarFinal(sc.nextInt());
                    break;

                case 3:
                    System.out.print("Dato: ");
                    int datoPos = sc.nextInt();
                    System.out.print("Posición: ");
                    int pos = sc.nextInt();
                    lista.insertarPosicion(datoPos, pos);
                    break;

                case 4:
                    lista.eliminarInicio();
                    break;

                case 5:
                    lista.eliminarFinal();
                    break;

                case 6:
                    System.out.print("Posición a eliminar: ");
                    lista.eliminarPosicion(sc.nextInt());
                    break;

                case 7:
                    lista.mostrar();
                    break;

                case 8:
                    System.out.println("Saliendo...");
                    break;

                default:
                    System.out.println("Opción no válida");
            }

        } while (opcion != 8);

        sc.close();
    }
}

File: listacirculardoble/listaCirculardoble.js
const prompt = require('prompt-sync')({sigint: true});

class Nodo {
    constructor(dato){ this.dato = dato; this.sig = null; this.ant = null; }
}

class ListaCircularDoble {
    constructor(){ this.cabeza = null; }

    insertarInicio(dato){
        const nuevo = new Nodo(dato);
        if(!this.cabeza){ this.cabeza = nuevo; nuevo.sig = nuevo; nuevo.ant = nuevo; return; }
        const ultimo = this.cabeza.ant;
        nuevo.sig = this.cabeza; nuevo.ant = ultimo;
        ultimo.sig = nuevo; this.cabeza.ant = nuevo;
        this.cabeza = nuevo;
    }

    insertarFinal(dato){
        const nuevo = new Nodo(dato);
        if(!this.cabeza){ this.cabeza = nuevo; nuevo.sig = nuevo; nuevo.ant = nuevo; return; }
        const ultimo = this.cabeza.ant;
        ultimo.sig = nuevo; nuevo.ant = ultimo;
        nuevo.sig = this.cabeza; this.cabeza.ant = nuevo;
    }

    insertarPosicion(dato, pos){
        if(pos === 0){ this.insertarInicio(dato); return; }
        if(!this.cabeza){ console.log("Lista vacía"); return; }
        let temp = this.cabeza;
        for(let i=0;i<pos-1 && temp.sig !== this.cabeza;i++) temp = temp.sig;
        const nuevo = new Nodo(dato);
        const siguiente = temp.sig;
        nuevo.sig = siguiente; nuevo.ant = temp;
        temp.sig = nuevo; siguiente.ant = nuevo;
    }

    eliminarInicio(){
        if(!this.cabeza) return;
        if(this.cabeza.sig === this.cabeza){ this.cabeza = null; return; }
        const ultimo = this.cabeza.ant;
        this.cabeza = this.cabeza.sig;
        this.cabeza.ant = ultimo;
        ultimo.sig = this.cabeza;
    }

    eliminarFinal(){
        if(!this.cabeza) return;
        if(this.cabeza.sig === this.cabeza){ this.cabeza = null; return; }
        const ultimo = this.cabeza.ant;
        const nuevoUltimo = ultimo.ant;
        nuevoUltimo.sig = this.cabeza;
        this.cabeza.ant = nuevoUltimo;
    }

    eliminarPosicion(pos){
        if(!this.cabeza) return;
        if(pos === 0){ this.eliminarInicio(); return; }
        let temp = this.cabeza;
        for(let i=0;i<pos;i++){
            temp = temp.sig;
            if(temp === this.cabeza) return;
        }
        const anterior = temp.ant;
        const siguiente = temp.sig;
        anterior.sig = siguiente; siguiente.ant = anterior;
        if(temp === this.cabeza) this.cabeza = siguiente;
    }

    mostrar(){
        if(!this.cabeza){ console.log("Lista vacía"); return; }
        let temp = this.cabeza;
        do {
            process.stdout.write(temp.dato + " <-> ");
            temp = temp.sig;
        } while(temp !== this.cabeza);
        console.log("(regresa)");
    }
}

// Menú
const lista = new ListaCircularDoble();
let opcion;
do {
    console.log("\n=== MENÚ LISTA CIRCULAR DOBLE (JS) ===");
    console.log("1. Insertar al inicio\n2. Insertar al final\n3. Insertar en posición");
    console.log("4. Eliminar al inicio\n5. Eliminar al final\n6. Eliminar en posición");
    console.log("7. Mostrar lista\n8. Salir");
    opcion = parseInt(prompt("Opción: "));
    switch(opcion){
        case 1:
            lista.insertarInicio(parseInt(prompt("Dato: ")));
            break;
        case 2:
            lista.insertarFinal(parseInt(prompt("Dato: ")));
            break;
        case 3:
            lista.insertarPosicion(parseInt(prompt("Dato: ")), parseInt(prompt("Posición: ")));
            break;
        case 4:
            lista.eliminarInicio();
            break;
        case 5:
            lista.eliminarFinal();
            break;
        case 6:
            lista.eliminarPosicion(parseInt(prompt("Posición a eliminar: ")));
            break;
        case 7:
            lista.mostrar();
            break;
        case 8:
            console.log("Saliendo...");
            break;
        default:
            console.log("Opción no válida");
    }
} while(opcion !== 8);
